---
title: "Tutorial phyloseq"
output: github_document
---
A recharger à chaque fois
```{r}
library(dada2)
library(phyloseq)
library(DECIPHER)
library(phangorn)
library(ggplot2)
library(gridExtra)
```


```{r}
#definir la variable du repertoire
miseq_path <- "/home/rstudio/MiSeq_SOP" 
list.files(miseq_path)
```

```{r}
#Le tri assure la lecture forward et reverse dans le meme ordre 
fnFs <- sort(list.files(miseq_path, pattern="_R1_001.fastq"))
fnRs <- sort(list.files(miseq_path, pattern="_R2_001.fastq"))
# Extraction des noms des echantillons en supposant que les noms de fichiers ont le format : SAMPLENAME_XXX.fastq
sampleNames <- sapply(strsplit(fnFs, "_"), `[`, 1)
# Indiquer le chemin complet vers les fnFs (forward) et fnRs (reverse) 
fnFs <- file.path(miseq_path, fnFs)
fnRs <- file.path(miseq_path, fnRs)
fnRs[1:3]
```

```{r}
#profil de qualité des lectures Forward
plotQualityProfile(fnFs[1:2])

#bonne qualité
```

```{r}
#profil de qualité des lectures Reverse
plotQualityProfile(fnRs[1:2])

#mauvaise qualité (tronquer les séquences à la position 160)
```

```{r}
# les fichiers filtrés sont placés dans le sous-répertoire filtered/.
filt_path <- file.path(miseq_path, "filtered") 
if(!file_test("-d", filt_path)) dir.create(filt_path)
filtFs <- file.path(filt_path, paste0(sampleNames, "_F_filt.fastq.gz"))
filtRs <- file.path(filt_path, paste0(sampleNames, "_R_filt.fastq.gz"))
```

```{r}
#Définition des paramètres de filtration
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(240,160),
              maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE) # On Windows set multithread=FALSE
head(out)
```

```{r}
#déréplication : élimine les comparaisons redondantes
derepFs <- derepFastq(filtFs, verbose=TRUE)
derepRs <- derepFastq(filtRs, verbose=TRUE)
#Noms des échantillons attribués aux objets derep
names(derepFs) <- sampleNames
names(derepRs) <- sampleNames
```

```{r}
#utilisation de la fonction learnErrors : distinction des erreurs de séquencage avec les variations biologiques réelles
errF <- learnErrors(filtFs, multithread=TRUE)
errR <- learnErrors(filtFs, multithread=TRUE)
```

```{r}
#estimation du taux d'erreurs pour la lecture Forward 
plotErrors(errF)
```
```{r}
#estimation du taux d'erreurs pour la lecture Reverse
plotErrors(errR)
```
```{r}
#Inférence à partir de séquençage groupé, à partir de tous les échantillons. Forward
dadaFs <- dada(derepFs, err=errF, multithread=TRUE)
```

```{r}
#Reverse
dadaRs <- dada(derepRs, err=errR, multithread=TRUE)
```

```{r}
#Inspection de l'objet
dadaFs[[1]]
```

```{r}
#Fusion des séquences Forward et Reverse inférées (supression des séquences appariées qui ne se chevauchent pas)
mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs)
```

```{r}
#Construction d'un tableau de séquences
seqtabAll <- makeSequenceTable(mergers[!grepl("Mock", names(mergers))])
table(nchar(getSequences(seqtabAll)))
```

```{r}
#suppression des chimères
seqtabNoC <- removeBimeraDenovo(seqtabAll)
```

```{bash}
wget https://zenodo.org/record/4310151/files/rdp_train_set_18.fa.gz
```

```{r}
#Attribution d'une taxonomie
fastaRef <- "./rdp_train_set_18.fa.gz"
taxTab <- assignTaxonomy(seqtabNoC, refFasta = fastaRef, multithread=TRUE)
unname(head(taxTab))
```

```{r}
#combiner des données en objet phyloseq #1
samdf <- read.csv("https://raw.githubusercontent.com/spholmes/F1000_workflow/master/data/MIMARKS_Data_combined.csv",header=TRUE)
samdf$SampleID <- paste0(gsub("00", "", samdf$host_subject_id), "D", samdf$age-21)
samdf <- samdf[!duplicated(samdf$SampleID),] #Suppression des duplication pour les lectures Reverse
rownames(seqtabAll) <- gsub("124", "125", rownames(seqtabAll)) #Correction de la divergence 
all(rownames(seqtabAll) %in% samdf$SampleID) # TRUE
print(samdf$SampleID)
```
```{r}
#alignement de séquences multiples
seqs <- getSequences(seqtabNoC)
names(seqs) <- seqs # This propagates to the tip labels of the tree
alignment <- AlignSeqs(DNAStringSet(seqs), anchor=NA,verbose=FALSE)
```

```{r}
#Construction d'un arbre phylogénétique
phangAlign <- phyDat(as(alignment, "matrix"), type="DNA")
dm <- dist.ml(phangAlign)
treeNJ <- NJ(dm) # Note, tip order != sequence order
fit = pml(treeNJ, data=phangAlign)
fitGTR <- update(fit, k=4, inv=0.2)
fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
        rearrangement = "stochastic", control = pml.control(trace = 0))
detach("package:phangorn", unload=TRUE)
```

```{r}
#combiner des données en objet phyloseq #2
rownames(samdf) <- samdf$SampleID
keep.cols <- c("collection_date", "biome", "target_gene", "target_subfragment",
"host_common_name", "host_subject_id", "age", "sex", "body_product", "tot_mass",
"diet", "family_relationship", "genotype", "SampleID") 
samdf <- samdf[rownames(seqtabAll), keep.cols]
```

```{r}
#chargement des données dans phyloseq
ps <- phyloseq(otu_table(seqtabNoC, taxa_are_rows=FALSE), 
               sample_data(samdf), 
               tax_table(taxTab),phy_tree(fitGTR$tree))
ps <- prune_samples(sample_names(ps) != "Mock", ps) 
ps
```
```{r}
ps_connect <-url("https://raw.githubusercontent.com/spholmes/F1000_workflow/master/data/ps.rds")
ps = readRDS(ps_connect)
ps
```

```{r}
test <- ps@phy_tree
plot_tree(test)
```

```{r}
ps <- subset_taxa(ps, !is.na(Phylum) & !Phylum %in% c("", "uncharacterized"))
```

```{r}
# Calculer la prévalence de chaque caractéristique, la stocker dans data.frame.
prevdf = apply(X = otu_table(ps),
               MARGIN = ifelse(taxa_are_rows(ps), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})
#Ajoutez la taxonomie et le nombre total de lectures à ce cadre de données.
prevdf = data.frame(Prevalence = prevdf,
                    TotalAbundance = taxa_sums(ps),
                    tax_table(ps))
```

```{r}
#Filtrage taxonomique
plyr::ddply(prevdf, "Phylum", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})
```

```{r}
# Définir les phyla à filtrer 
filterPhyla = c("Fusobacteria", "Deinococcus-Thermus")
# Filtrer les entrées dont le phylum n'est pas identifié. 
ps1 = subset_taxa(ps, !Phylum %in% filterPhyla)
ps1
```
```{r}
#Filtrage non supervisée de la prévalence (prevalence = nombre d'echantillons dans lequel apparaissent les taxas)
#Sous-ensemble des autres phyla
prevdf1 = subset(prevdf, Phylum %in% get_taxa_unique(ps1, "Phylum"))
ggplot(prevdf1, aes(TotalAbundance, Prevalence / nsamples(ps),color=Phylum)) +
  # Inclure une estimation pour le paramètre
  geom_hline(yintercept = 0.05, alpha = 0.5, linetype = 2) +  geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() +  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~Phylum) + theme(legend.position="none")
```
```{r}
# Définir 5% de tous les échantillons comme seuil de prévalence.
prevalenceThreshold = 0.05 * nsamples(ps)
prevalenceThreshold
```

```{r}
# Exécuter le filtre de prévalence, en utilisant la fonction `prune_taxa()`.
keepTaxa = rownames(prevdf1)[(prevdf1$Prevalence >= prevalenceThreshold)]
ps2 = prune_taxa(keepTaxa, ps)
```

```{r}
# Combien de genres seraient présents après le filtrage ?
length(get_taxa_unique(ps2, taxonomic.rank = "Genus"))
```

```{r}
#taux d'agglomération 
ps3 = tax_glom(ps2, "Genus", NArm = TRUE)
```

```{r}
#Définir unee hauteur d'arbre correspondant à la distance phylogénétique entre les caractéristiques qui devraient définir leur regroupement
h1 = 0.4
ps4 = tip_glom(ps2, h = h1)
```

```{r}
#plot_tree compare les données originales non filtrées, l'arbre après agglomération taxonomique, et l'arbre après agglomération phylogénétique
multiPlotTitleTextSize = 15
p2tree = plot_tree(ps2, method = "treeonly",
                   ladderize = "left",
                   title = "Before Agglomeration") +
  theme(plot.title = element_text(size = multiPlotTitleTextSize))
p3tree = plot_tree(ps3, method = "treeonly",
                   ladderize = "left", title = "By Genus") +
  theme(plot.title = element_text(size = multiPlotTitleTextSize))
p4tree = plot_tree(ps4, method = "treeonly",
                   ladderize = "left", title = "By Height") +
  theme(plot.title = element_text(size = multiPlotTitleTextSize))
```

```{r}
#grouper le valeurs ensemble
grid.arrange(nrow = 1, p2tree, p3tree, p4tree)

#differents types d'agglomération
```
```{r}
#abondance relative
plot_abundance = function(physeq,title = "",
                          Facet = "Order", Color = "Phylum"){
  #Sous-ensemble arbitraire, basé sur le phylum
  p1f = subset_taxa(physeq, Phylum %in% c("Firmicutes"))
  mphyseq = psmelt(p1f)
  mphyseq <- subset(mphyseq, Abundance > 0)
  ggplot(data = mphyseq, mapping = aes_string(x = "sex",y = "Abundance",
                              color = Color, fill = Color)) +
    geom_violin(fill = NA) +
    geom_point(size = 1, alpha = 0.3,
               position = position_jitter(width = 0.3)) +
    facet_wrap(facets = Facet) + scale_y_log10()+
    theme(legend.position="none")
}
```

```{r}
# Transformer en abondance relative. Sauvegarder comme nouvel objet.
ps3ra = transform_sample_counts(ps3, function(x){x / sum(x)})
```

```{r}
#comparaison des abondances originales et des abondances relatives 
plotBefore = plot_abundance(ps3,"")
plotAfter = plot_abundance(ps3ra,"")
grid.arrange(nrow = 2,  plotBefore, plotAfter)
```

```{r}
#Attribution d'un rang taxonomique plus précis à l'argument de la fonction définie
psOrd = subset_taxa(ps3ra, Order == "Lactobacillales")
plot_abundance(psOrd, Facet = "Genus", Color = NULL)
```

```{r}
#installation de packages 

.cran_packages <- c( "shiny","miniUI", "caret", "pls", "e1071", "ggplot2", "randomForest", "dplyr", "ggrepel", "nlme", "devtools",
                  "reshape2", "PMA", "structSSI", "ade4",
                  "ggnetwork", "intergraph", "scales")
.github_packages <- c("jfukuyama/phyloseqGraphTest")
.bioc_packages <- c("genefilter", "impute")
# Install CRAN packages (if not already installed)
.inst <- .cran_packages %in% installed.packages()
if (any(!.inst)){
  install.packages(.cran_packages[!.inst],repos = "http://cran.rstudio.com/")
}
.inst <- .github_packages %in% installed.packages()
if (any(!.inst)){
  devtools::install_github(.github_packages[!.inst])
}

.inst <- .bioc_packages %in% installed.packages()
if(any(!.inst)){
  source("http://bioconductor.org/biocLite.R")
  biocLite(.bioc_packages[!.inst])
}
```

```{r}
library("shiny")
library("miniUI")
library("ggplot2")
library("randomForest")
library("dplyr")
library("caret")
library("pls")
library("e1071")
library("ggrepel")
library("nlme")
library("devtools")
library("reshape2")
library("PMA")
#library("structSSI")
library("ade4")
library("ggnetwork")
library("intergraph")
library("scales")
library("phyloseqGraphTest")
library("genefilter")
library("impute")
```

```{r}
#histogramme des groupes d'âges
qplot(sample_data(ps)$age, geom = "histogram",binwidth=20) + xlab("age")
```

```{r}
#Histogrammes comparant les données brutes et les comptages enregistrés par échantillon
qplot(log10(rowSums(otu_table(ps))),binwidth=0.2) +
  xlab("Logged counts-per-sample")
```
```{r}
#Exploration de l'ordination avec les abondances en log
sample_data(ps)$age_binned <- cut(sample_data(ps)$age,
                          breaks = c(0, 100, 200, 400))
levels(sample_data(ps)$age_binned) <- list(Young100="(0,100]", Mid100to200="(100,200]", Old200="(200,400]")
sample_data(ps)$family_relationship=gsub(" ","",sample_data(ps)$family_relationship)
pslog <- transform_sample_counts(ps, function(x) log(1 + x))
out.wuf.log <- ordinate(pslog, method = "MDS", distance = "wunifrac")
evals <- out.wuf.log$values$Eigenvalues
plot_ordination(pslog, out.wuf.log, color = "age_binned") +
  labs(col = "Binned Age") +
  coord_fixed(sqrt(evals[2] / evals[1]))
```

```{r}
rel_abund <- t(apply(otu_table(ps), 1, function(x) x / sum(x)))
qplot(rel_abund[, 12], geom = "histogram",binwidth=0.05) +
  xlab("Relative abundance")
```
```{r}
#Différentes projections d'ordination en effectuant d'abord un PCoA en utilisant la dissimilarité Bray-Curtis.
outliers <- c("F5D165", "F6D165", "M3D175", "M4D175", "M5D175", "M6D175")
ps <- prune_samples(!(sample_names(ps) %in% outliers), ps)
which(!rowSums(otu_table(ps)) > 1000)
ps <- prune_samples(rowSums(otu_table(ps)) > 1000, ps)
pslog <- transform_sample_counts(ps, function(x) log(1 + x))
out.pcoa.log <- ordinate(pslog,  method = "MDS", distance = "bray")
evals <- out.pcoa.log$values[,1]
plot_ordination(pslog, out.pcoa.log, color = "age_binned",
                  shape = "family_relationship") +
  labs(col = "Binned Age", shape = "Litter")+
  coord_fixed(sqrt(evals[2] / evals[1]))
```

 
```{r}
#Analyse de la DPCoA (Calculate Double Principle Coordinate Analysis) : méthode d'ordination phylogénétique, qui fournit une représentation biplot des échantillons et des catégories taxonomiques.
out.dpcoa.log <- ordinate(pslog, method = "DPCoA")
evals <- out.dpcoa.log$eig
plot_ordination(pslog, out.dpcoa.log, color = "age_binned", label= "SampleID",
                  shape = "family_relationship") +
  labs(col = "Binned Age", shape = "Litter")+
  coord_fixed(sqrt(evals[2] / evals[1]))
```

```{r}
plot_ordination(pslog, out.dpcoa.log, type = "species", color = "Phylum") +
  coord_fixed(sqrt(evals[2] / evals[1]))
```
```{r}
#Analyse des résultats de PCoA avec Unifrac pondéré
out.wuf.log <- ordinate(pslog, method = "PCoA", distance ="wunifrac")
evals <- out.wuf.log$values$Eigenvalues
plot_ordination(pslog, out.wuf.log, color = "age_binned",
                  shape = "family_relationship") +
  coord_fixed(sqrt(evals[2] / evals[1])) +
  labs(col = "Binned Age", shape = "Litter")
```
```{r}
#nouvelle matrice représentant les abondances par leurs rangs (microbe le moins abondant dans un échantillon est affecté au rang 1)
abund <- otu_table(pslog)
abund_ranks <- t(apply(abund, 1, rank))

#tous les microbes dont le rang est inférieur à un certain seuil sont mis à égalité à 1
abund_ranks <- abund_ranks - 329
abund_ranks[abund_ranks < 1] <- 1

library(dplyr)
library(reshape2)
abund_df <- melt(abund, value.name = "abund") %>%
  left_join(melt(abund_ranks, value.name = "rank"))
colnames(abund_df) <- c("sample", "seq", "abund", "rank")

abund_df <- melt(abund, value.name = "abund") %>%
  left_join(melt(abund_ranks, value.name = "rank"))
colnames(abund_df) <- c("sample", "seq", "abund", "rank")

sample_ix <- sample(1:nrow(abund_df), 8)
ggplot(abund_df %>%
         filter(sample %in% abund_df$sample[sample_ix])) +
  geom_point(aes(x = abund, y = rank, col = sample),
             position = position_jitter(width = 0.2), size = 1.5) +
  labs(x = "Abundance", y = "Thresholded rank") +
  scale_color_brewer(palette = "Set2")
```

```{r}
#ACP et étude du biplot résultant
library(ade4)
ranks_pca <- dudi.pca(abund_ranks, scannf = F, nf = 3)
row_scores <- data.frame(li = ranks_pca$li,
                         SampleID = rownames(abund_ranks))
col_scores <- data.frame(co = ranks_pca$co,
                         seq = colnames(abund_ranks))
tax <- tax_table(ps) %>%
  data.frame(stringsAsFactors = FALSE)
tax$seq <- rownames(tax)
main_orders <- c("Clostridiales", "Bacteroidales", "Lactobacillales",
                 "Coriobacteriales")
tax$Order[!(tax$Order %in% main_orders)] <- "Other"
tax$Order <- factor(tax$Order, levels = c(main_orders, "Other"))
tax$otu_id <- seq_len(ncol(otu_table(ps)))
row_scores <- row_scores %>%
  left_join(sample_data(pslog))
col_scores <- col_scores %>%
  left_join(tax)

evals_prop <- 100 * (ranks_pca$eig / sum(ranks_pca$eig))
ggplot() +
  geom_point(data = row_scores, aes(x = li.Axis1, y = li.Axis2), shape = 2) +
  geom_point(data = col_scores, aes(x = 25 * co.Comp1, y = 25 * co.Comp2, col = Order),
             size = .3, alpha = 0.6) +
  scale_color_brewer(palette = "Set2") +
  facet_grid(~ age_binned) +
  guides(col = guide_legend(override.aes = list(size = 3))) +
  labs(x = sprintf("Axis1 [%s%% variance]", round(evals_prop[1], 2)),
       y = sprintf("Axis2 [%s%% variance]", round(evals_prop[2], 2))) +
  coord_fixed(sqrt(ranks_pca$eig[2] / ranks_pca$eig[1])) +
  theme(panel.border = element_rect(color = "#787878", fill = alpha("white", 0)))
```
```{r}
#analyse de la correspondance canonique (CCpnA)
ps_ccpna <- ordinate(pslog, "CCA", formula = pslog ~ age_binned + family_relationship)

library(ggrepel)
ps_scores <- vegan::scores(ps_ccpna)
sites <- data.frame(ps_scores$sites)
sites$SampleID <- rownames(sites)
sites <- sites %>%
  left_join(sample_data(ps))

species <- data.frame(ps_scores$species)
species$otu_id <- seq_along(colnames(otu_table(ps)))
species <- species %>%
  left_join(tax)
evals_prop <- 100 * ps_ccpna$CCA$eig[1:2] / sum(ps_ccpna$CA$eig)
ggplot() +
  geom_point(data = sites, aes(x = CCA1, y = CCA2), shape = 2, alpha = 0.5) +
  geom_point(data = species, aes(x = CCA1, y = CCA2, col = Order), size = 0.5) +
  geom_text_repel(data = species %>% filter(CCA2 < -2),
                    aes(x = CCA1, y = CCA2, label = otu_id),
            size = 1.5, segment.size = 0.1) +
  facet_grid(. ~ family_relationship) +
  guides(col = guide_legend(override.aes = list(size = 3))) +
  labs(x = sprintf("Axis1 [%s%% variance]", round(evals_prop[1], 2)),
        y = sprintf("Axis2 [%s%% variance]", round(evals_prop[2], 2))) +
  scale_color_brewer(palette = "Set2") +
  coord_fixed(sqrt(ps_ccpna$CCA$eig[2] / ps_ccpna$CCA$eig[1])*0.45   ) +
  theme(panel.border = element_rect(color = "#787878", fill = alpha("white", 0)))
```

 